### PR TITLE
17dep distances@main

### LINKED ISSUE #17: propose direct dependencies
Add `update_dependencies` from the dependency graph (via `DESCRIPTION`) so that for every package we can list the complete distance 1 downstream & upstream packages. Please also add a distance to the dependency table

### PR DESCRIPTION
**Pull request by nikolas-burkoff**
Fix #17 

1) Uses DESCRIPTION file based dependencies 
2) Adds distance to dependency_table (using dependency graph from 1) - thanks @pawelru for the function
image
3) Adds `update_stage_dep_yaml_with_direct_deps` function

I've tried to make smallest number of changes to code base.

Still some tidying up etc. to do

### CODE REVIEW

**Comment by nikolas-burkoff** [comment]
Note the dependency graph for tern shown above is different to the old one from [here](https://github.com/insightsengineering/tern/runs/3190301432?check_suite_focus=true)

image

which looks wrong to me as for example hermes has tern in its description file   https://github.com/insightsengineering/hermes/blob/f5151ed62db2ee27170c0b41e3bd107aa397fd22/DESCRIPTION#L45

File: R/caching.R [hunk 1]
```diff
@@ -114,6 +114,19 @@ get_deps_info <- function(repo_dir) {
   }
 }
 
+
+get_description_deps <- function(repo_dir) {
```
Comment (maximilianmordig) [inline R/caching.R:118]: move these two functions to `utils.R`

File: R/dependencies.R [hunk 2]
```diff
@@ -698,5 +704,32 @@ get_local_pkgs_from_config <- function() {
 
 
 
+#' TODO roxygen
+#' @export
+update_stage_dep_yaml_with_direct_deps <- function(project = ".", feature = NULL,
```
Comment (maximilianmordig) [inline R/dependencies.R:709]: name too long and difficult to read, inconsistent with above
maybe `update_stageddeps_yaml_with_direct_deps`

File: R/caching.R [hunk 3]
```diff
@@ -168,10 +181,13 @@ rec_checkout_repos <- function(repos_to_process, feature, direction = c("upstrea
...
 
-  # only one of them may be filled depending on the direction
-  # should not be c() since assigned to list below removes element otherwise
-  upstream_deps_graph <- list()
-  downstream_deps_graph <- list()
+  hashed_processed_repos <- character(0)
+
+  # named list of packages dependencies (depends, suggest, imports)
+  # from DESCRIPTION file of package
+  r_description_file_deps <- list()
```
Comment (maximilianmordig) [inline R/caching.R:188]: remove `r_` prefix

File: R/caching.R [hunk 4]
```diff
@@ -198,31 +214,46 @@ rec_checkout_repos <- function(repos_to_process, feature, direction = c("upstrea
       )
     }
 
+    r_description_file_deps[[hashed_repo_and_host]] <- get_description_deps(repo_dir)
+    r_package_name[hashed_repo_and_host] <- get_package_name(repo_dir)
```
[the remark this answers is not in the document]
Comment (nikolas-burkoff) [reply R/caching.R:218]: This is (or at least should be) a named vector not a list (much easier to manipulate) which only needs `[]` not `[[]]`

File: R/caching.R [hunk 5]
```diff
@@ -168,10 +181,13 @@ rec_checkout_repos <- function(repos_to_process, feature, direction = c("upstrea
...
-  # should not be c() since assigned to list below removes element otherwise
-  upstream_deps_graph <- list()
-  downstream_deps_graph <- list()
+  hashed_processed_repos <- character(0)
+
+  # named list of packages dependencies (depends, suggest, imports)
+  # from DESCRIPTION file of package
+  r_description_file_deps <- list()
+  # Named vector mapping hashed_repo_and_host to package name
+  r_package_name <- c()
```
Comment (maximilianmordig) [inline R/caching.R:190]: find a better name

File: R/caching.R [hunk 6]
```diff
@@ -168,10 +181,13 @@ rec_checkout_repos <- function(repos_to_process, feature, direction = c("upstrea
   hashed_repos_to_process <- vapply(repos_to_process, hash_repo_and_host, character(1))
   rm(repos_to_process)
 
-  # only one of them may be filled depending on the direction
-  # should not be c() since assigned to list below removes element otherwise
-  upstream_deps_graph <- list()
-  downstream_deps_graph <- list()
+  hashed_processed_repos <- character(0)
```
Comment (maximilianmordig) [inline R/caching.R:184]: update function doc, what exactly is returned, distincttion btwenn dependency graph from yaml and description file

File: R/dependencies.R [hunk 7]
```diff
@@ -698,5 +704,32 @@ get_local_pkgs_from_config <- function() {
...
+  upstream_repos <- dep_table[dep_table$type == "upstream" & dep_table$distance == 1, c("repo", "host")]
+  upstream_repos = setNames(apply(upstream_repos, 1, FUN = function(row) row = list(repo=row["repo"], host = row["host"])), NULL)
+
+  downstream_repos <- dep_table[dep_table$type == "downstream" & dep_table$distance == 1, c("repo", "host")]
+  downstream_repos <- setNames(apply(downstream_repos, 1, FUN = function(row) row = list(repo=row["repo"], host = row["host"])), NULL)
+
+  yaml::write_yaml(
+    list(current_repo = current_repo,
+      upstream_repos = upstream_repos,
+      downstream_repos = downstream_repos), file = file.path(project, "staged_dependencies.yaml"))
```
Comment (maximilianmordig) [inline R/dependencies.R:733]: hardcoded, use `STAGEDDEPS_FILENAME` instead

File: R/caching.R [hunk 8]
```diff
@@ -114,6 +114,19 @@ get_deps_info <- function(repo_dir) {
...
+  dsc <- desc::desc(fdesc)
+  return(dsc$get_deps()$package)
+}
+
+get_package_name <- function(repo_dir) {
+  fdesc <- file.path(repo_dir, "DESCRIPTION")
+  dsc <- desc::desc(fdesc)
+  return(unname(dsc$get("Package")))
+}
+
```
Comment (maximilianmordig) [inline R/caching.R:129]: on line 136: hardcoded yaml name, use `STAGEDDEPS_FILENAME` instead

File: R/graph_methods.R [hunk 9]
```diff
@@ -62,6 +62,28 @@ topological_sort <- function(graph) {
...
+      return()
+    }
+    c(
+      lapply(parents_to_children[[node]], list, distance + 1L),
+      unlist(lapply(parents_to_children[[node]],
+                    internal_recursive_fun, parents_to_children = parents_to_children,
+                    distance = distance + 1L), recursive = FALSE)
+    )
+  }
+  internal_recursive_fun(parents_to_children, node) %>%
```
Comment (maximilianmordig) [inline R/graph_methods.R:79]: this is a bad name for the function
then, you can just call it `f`, that is equally informative (and maybe make a comment that it is recursively called)

**Comment by maximilianmordig** [comment]
This PR does not convince me, the solution is hacky. I would keep the package dependency parsing outside of `checkout_repo` and do it in a separate loop once the graph based on the yamls has been obtained. In other words, I would revert the changes to `checkout_repo`.

File: R/dependencies.R [hunk 10]
```diff
@@ -482,25 +482,31 @@ dependency_structure <- function(project = ".", feature = NULL,
...
     ),
     cbind_handle_empty(
-      data.frame(unhash_repo_and_host(hashed_downstream_nodes), stringsAsFactors = FALSE), type = "downstream"
+      data.frame(unhash_repo_and_host(hashed_downstream_nodes$id),
+                 distance = hashed_downstream_nodes$distance,
+                 stringsAsFactors = FALSE), type = "downstream"
     ),
     cbind_handle_empty(
-      data.frame(unhash_repo_and_host(hashed_remaining_nodes), stringsAsFactors = FALSE), type = "other"
+      data.frame(unhash_repo_and_host(hashed_remaining_nodes),
```
Comment (maximilianmordig) [inline R/dependencies.R:508]: does  not work when `hashed_remaining_nodes` is of length 0

Comment (maximilianmordig) [reply R/dependencies.R:508]:
```
cbind_handle_empty(
      data.frame(unhash_repo_and_host(hashed_remaining_nodes),
                 stringsAsFactors = FALSE), distance = as.numeric(NA), type = "other"
    )
```
works

**Review by maximilianmordig — CHANGES_REQUESTED** [review]
see my comment above

**Comment by maximilianmordig** [comment]
Also fix all warnings from R Cmd check

File: R/caching.R (hunk 6 above)
Comment (maximilianmordig) [reply R/caching.R:184]: I find this function confusing and would revert it to its old state. Then get the package dependencies (according to DESCRIPTKION) in another loop.

File: R/graph_methods.R (hunk 9 above)
Comment (nikolas-burkoff) [reply R/graph_methods.R:79]: blame @pawelru  :)

**Comment by nikolas-burkoff** [comment]
Cool thanks for review - lost of good points I'll address ASAP (I wanted a draft out there early to speed up the feedback cycle) 

> This PR does not convince me, the solution is hacky. I would keep the package dependency parsing outside of checkout_repo and do it in a separate loop once the graph based on the yamls has been obtained. In other words, I would revert the changes to checkout_repo.

Happy to move the calculation of dependency parsing outside of checkout_repo I think that makes sense.

But checkout_repo must be changed to just return the list of internal packages as if it is left alone then it returns a possibly incorrect dependency graph (and definitely incorrect for NEST as tern is an upstream dep of Hermes and that is missed in the yaml files) based on yaml which we then throw away and replace with the correct graph from the DESCRIPTION file.

So this is what I propose:

```
# this is the output of a new version of checkout_repo which returns a list of "repo @ host" packages
# which will be upstream/downstream or both from current repo 
internal_dependencies <- get_all_internal_dependencies(...)
# this returns the true dependency graph given the list of internal packages
deps <- get_dependency_grap(internal_dependencies)   
```

File: R/caching.R (hunk 1 above)
Comment (nikolas-burkoff) [reply R/caching.R:118]: functions no longer needed after refator

File: R/caching.R (hunk 3 above)
Comment (nikolas-burkoff) [reply R/caching.R:188]: no longer needed after refactor

File: R/caching.R (hunk 5 above)
Comment (nikolas-burkoff) [reply R/caching.R:190]: not used now after refactor

File: R/caching.R (hunk 6 above)
Comment (nikolas-burkoff) [reply R/caching.R:184]: no longer relevant after refactor

File: R/graph_methods.R [hunk 11]
```diff
@@ -62,6 +62,28 @@ topological_sort <- function(graph) {
   }
 }
 
+# get the descendants (all children) of node, given list mapping parent to children
+# and their distances
+get_descendants_distance <- function(parents_to_children, node) {
+  internal_recursive_fun <- function(parents_to_children, node, distance = 0L) {
```
[the remark this answers is not in the document]
Comment (nikolas-burkoff) [reply R/graph_methods.R:68]: full unit tests of this function required - will add

File: R/dependencies.R (hunk 10 above)
Comment (nikolas-burkoff) [reply R/dependencies.R:508]: thanks 👍

File: R/graph_methods.R (hunk 9 above)
Comment (nikolas-burkoff) [reply R/graph_methods.R]: Changed to an iterative not recursive function

File: R/graph_methods.R (hunk 11 above)
[the remark this answers is not in the document]
Comment (nikolas-burkoff) [reply R/graph_methods.R]: tests added

File: R/dependencies.R [hunk 12]
```diff
@@ -697,6 +718,94 @@ get_local_pkgs_from_config <- function() {
...
+                                                    verbose = 0){
+  dep_table <- dependency_structure(
+    project = project,
+    feature = feature,
+    local_repos = local_repos,
+    return_table_only = TRUE,
+    verbose = verbose
+  )
 
+  current_repo <- list("repo" = dep_table[dep_table$type == "current",]$repo,
```
Comment (nikolas-burkoff) [inline R/dependencies.R:797]: I think we should refactor this so we can test it and add tests (or if we don't do it now we add it as an issue) - if written like this:

```
update_stageddeps_yaml_with_direct_deps <- function(){

  dep_table <- get_dependency_table() 
  yaml_string <- get_yaml_from_dependency_table(dep_table) #this returns a string 
  write yaml_string to file = file.path(project, STAGEDDEPS_FILENAME) #don't need write_yaml
}
```

Then we can unit test get_yaml_from_dependency_table by plugging in different (made up) dependency tables and check the yaml that comes out is correct

File: R/caching.R [hunk 13]
```diff
@@ -152,10 +153,13 @@ get_hashed_repo_to_dir_mapping <- function(local_repos) {
...
 #'   2: includes git checkout infos)
 #'
-#' @return the upstream and/or downstream dependency graphs
-rec_checkout_repos <- function(repos_to_process, feature, direction = c("upstream"),
-                               local_repos = get_local_pkgs_from_config(),
-                               verbose = 0) {
+#' @return A list, on entry per checkout out repository whose name is the name of the
+#'   repo in the form x @@ host and whose value is the directory to
+#'   which it has been checked out into
+get_internal_dependencies <- function(repos_to_process, feature,
```
Comment (maximilianmordig) [inline R/caching.R:159]: I would rename this back to `rec_checkout_repos` because it has a significant side effect.

Comment (maximilianmordig) [reply R/caching.R:159]: renamed to `rec_checkout_internal_deps`

File: R/dependencies.R (hunk 12 above)
Comment (maximilianmordig) [reply R/dependencies.R:797]: done

**Review by maximilianmordig — APPROVED** [review]

**Pull request merged**